Add support for pluralization string extraction and use

jonathannewman opened the pull request:
This adds two new functions trun and trsn that support pluralization.  They mirror the
ngettext function in behavior, while keeping the convention of java formatting.

Examples and tests were updated as well.

jonathannewman commented:
@lutter @libbymo PR to add pluralization support.  I don't know that the German examples are correct, as they came from Google Translate.  

Please let me know what you think.

libbymo commented on src/puppetlabs/i18n/main.clj:
```diff
@@ -17,10 +17,24 @@
+      ; first item is number of programmers, second is number of months
+      (println (str (i18n/tru "It took {0} {1} to implement this"
+                              (i18n/trun "one programmer" "{0} programmers" nprog)
+                              (i18n/trun "one month" "{0} months" nmonths))))))
```
@LarissaLane This is the example we were talking about the other day in the meeting, of say two items that may need pluralization in a single sentence.  I think this is a good (if potentially the only?) way to solve this problem.  Is this clear to you (and or other translators?) Would it require further comment/context in an actual application?

lutter commented on README.md:
```diff
@@ -22,19 +22,29 @@ for example log messages, and `tru` for messages that will be shown to the
+
+    (println (trs "It takes {0} {1} to change a light bulb"
+                  (trsn "one software engineer" "{0} software engineers" 3)
+                  (trsn "one hour" "{0} hours}" 9)))
```
I would hugely prefer if the example in the README addressed a more straightforward pluralization issue instead of immediately diving into double-pluralization. Something as boring as `(trsn "We found one file" "We found {0} files" 7)` would allow the reader to focus on the basic usage of `trsn` better.

jonathannewman replied:
Sounds good to me.

lutter commented on src/puppetlabs/i18n/main.clj:
```diff
@@ -17,10 +17,24 @@
 
+  ;; Very simple plural system localization
+  (println (trs "{0} of beer on the wall."
+                (trsn "One bottle" "{0} bottles" 99)))
```
This should not interpolate text into other text. Instead it should be
```clojure
(println (trsn "One bottle of beer on the wall" "{0} bottles of beer on the wall" 99))
```

jonathannewman replied:
Great point.

lutter commented on locales/de.po:
```diff
@@ -18,14 +18,42 @@ msgstr ""
+#. Very simple plural system localization
+#: src//puppetlabs/i18n/main.clj:21
+msgid "{0} of beer on the wall."
+msgstr "{0} von Bier an der Wand."
```
`"{0} Bier an der Wand"`

lutter commented on locales/de.po:
```diff
@@ -18,14 +18,42 @@ msgstr ""
+#: src//puppetlabs/i18n/main.clj:22
+msgid "One bottle"
+msgid_plural "{0} bottles"
+msgstr[0] "eine Flasche"
```
`"Eine Flasche"`

lutter commented on locales/de.po:
```diff
@@ -18,14 +18,42 @@ msgstr ""
+msgstr[0] "ein Monat"
+msgstr[1] "{0} Monate"
+
+#: src//puppetlabs/i18n/main.clj:41
```
This example is where interpolating text into other text gets tricky. The correct sentences in German would be
```
Ein Programmierer brauchte einen Monat um das zu implementieren
Ein Programmierer brauchte 5 Monate um das zu implementieren
3 Programmierer brauchten einen Monat um das zu implementieren
3 Programmierer brauchten 5 Monate um das zu implementieren
```
While you can still structure this as `"{0} brauchte(n) {1} um das zu implementieren"`, pluralizing the sentence based on the number of programmers, the correct translation for `one month` and `{0} months` depends on the sentence this gets interpolated into (yay for cases); for example, if the sentence was `"We are going on vacation in {0}"` the translation for `"5 months"` would now be `"5 Monaten"` rather than `"5 Monate"` in the previous example (Dativ vs Akkusativ for the grammar geeks ;) ).

I am pretty sure that these sort of complications will be exceedingly rare in our code, and think the example should just be simplified to just have one thing in it that triggers pluralization.

pcarlisle commented:
The minimum unit of translation should be a sentence, it is impossible to consistently translate sentence fragments. We should probably write that somewhere as a guideline.

jonathannewman commented:
Updated: README, example code and tests.   All the cases that used pluralization now only contain single sentences.   I don't know that the German translations are correct.  :)

libbymo commented on test/puppetlabs/i18n/core_test.clj:
```diff
@@ -37,6 +47,22 @@
+    (with-user-locale eo (is (= one_bottle_eo (trun one_bottle n_bottles 1))))
+
+;; this isn't working (how should properties file be formatted?
+;;    (with-user-locale eo (is (= one_bottle_eo (trun one_bottle n_bottles 7))))
```
Hmm, I have no idea. @lutter ? Is testing pluralization possible with the "eo" properties file?

lutter replied:
I just uncommented that test and `lein test` passes for me with that test in it (lein says it ran 10 tests with 59 assertions)

lutter replied:
I just actually read the test; I am not sure how a properties file should be formatted to support pluralization.

jonathannewman replied:
I can remove the test.  I couldn't figure out how the properties file should be formatted for pluralization.   It just seemed worth calling out?

libbymo replied:
Well I think it passes, because it's not actually testing against a pluralized string? It would suggest that trun if no plural form is present just defaults to the original msgid, the test string in the eo properties file.  Which makes sense as the properties file is kind of dumb.  To actually test it, one would expect to do the following.
```
(def seven_bottle_eo "fake_seven_bottle")
(with-user-locale eo (is (= seven_bottle_eo (trun one_bottle n_bottles 7))))
```
I think the question is how would one set up a properties file to actually account for msgid[1]="fake_seven_bottle" as we have in a .po file.

I also question the validity or WHAT we're testing on line 59, as it appears to just fall back to the id string and ignore all pluralization specific parameters passed in.  Don't have any answers :) Just questions.

jonathannewman replied:
The problem may also result from not having a pluralization rule set for "eo".  I don't know what the default is.

jonathannewman commented:
I removed the properties files pluralization tests, and added a small note to the README where it references properties files that pluralization isn't currently supported in the properties files.

lutter commented on src/puppetlabs/i18n/core.clj:
```diff
@@ -201,16 +216,33 @@
 
+(defn translate-plural
+  "Translate a message into the given locale, interpolating as
+  needed. Messages are looked up in the resource bundle associated with the
```
The doc string should state explicitly that the `count` can always be interpolated as `"{0}"`

jonathannewman replied:
Great point.  I'll add that.

lutter commented:
Other than my one little nitpick, this looks great, and should get merged

libbymo commented:
Sweet, I can do the merging as soon as that last bit's in.  Thanks @jonathannewman @lutter !

libbymo commented:
LGTM!
![](http://i.giphy.com/FGlK61kCPosus.gif)
